### ISSUE TITLE
Фискализация для клиентов Robokassa и формирование второго чека

### ISSUE THREAD

**Issue by alka09** [yii-cms/yii2-robokassa#20]
Для фискализации необходимо передавать receipt в формате json, этот параметр включается в контрольную подпись запроса. Формирование второго чека - для этого нужно также в формате json сформировать запрос, который передается POST на URL: https://ws.roboxchange.com/RoboFiscal/Receipt/Attach
все это вот здесь - https://docs.robokassa.ru/?_ga=2.7372530.1175979893.1611648512-1040268572.1610786241#6865

**Comment by slava-o** [comment]
Для фискализации лучше сделать отдельные классы для передачи чеков в PaymentOptions (а не массива), ибо любая ошибка при формировании чека зафейлит его на стороне сервиса робочеков, а тк данные идут напрямую в налоговую из офд, то можно потом хапнуть немало проблем с бухгалтерией и налоговой отчетностью. В том числе чеки содержат много параметров налогообложения и разбираться в этом программисту проблематично для формирования массива. 

в моем форке от основной ветке за апрель 2020 есть формирование первого и второго чеков с классами

PS забыл сказать - в моем форке чеки проверены и работают на реальных деньгах (техподдержка ответила, что чеки работают только на настоящих деньгах и в тестовом режиме их проверить нельзя)

**Comment by c7v** [comment]
>на настоящих деньгах и в тестовом режиме их проверить нельзя)

Я так понимаю, что так или иначе с тестовыми чеками на реальных деньгах, придётся хапнуть проблем с ФНС?

**Comment by slava-o** [comment]
> Я так понимаю, что так или иначе с тестовыми чеками на реальных деньгах, придётся хапнуть проблем с ФНС?

@c7v верно, если в тестовых чеках передавать полную ерунду, то есть вероятность получить вопросы от ФНС :) 

в любом случае электронные чеки передаются ОФД (оператором фискальных данных) сразу в налоговую, как если например выбить чек на реальном кассовом аппарате в магазине

**Comment by c7v** [comment]
>полную ерунду, то есть вероятность получить вопросы от ФНС

@slava-o нашел решение! Если передать полную ерунду, то можно через СТП попросить отправить изменения и всё будет хорошо.

**Comment by slava-o** [comment]
> можно через СТП попросить отправить изменения и всё будет хорошо.

@c7v о спасибо, что сообщили! теперь буду знать, что чек можно скорректировать через поддержку, тк я не знал об этом

**Comment by c7v** [comment]
@gonimar у вас уже реализована фискализация? Если да, поделитесь кодом подтестирую.

**Comment by gonimar** [comment]
> @gonimar у вас уже реализована фискализация? Если да, поделитесь кодом подтестирую.

https://github.com/yii-cms/yii2-robokassa/blob/master/README.md#%D0%BF%D1%80%D0%B8%D0%BC%D0%B5%D1%80-%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D1%8B-%D0%B4%D0%BB%D1%8F-%D1%84%D0%B8%D1%81%D0%BA%D0%B0%D0%BB%D0%B8%D0%B7%D0%B0%D1%86%D0%B8%D0%B8
